### PR TITLE
Adds json struct tags to mirror yaml ones

### PR DESCRIPTION
**Pull request by owen-d**
This PR includes json struct tags to mirror the yaml ones in config objects. This allows downstream projects to consume these with a consistent key structure regardless of format. I'm hoping this isn't too controversial :).

For context: I'd like to consolidate these so Grafana can more easily present a JSON API over the Alertmanager structs and seamlessly transition json<>yaml. The lack of tag synchronization combined with the `<secret>` redaction in marshaling code makes this a bit more difficult than I'd like.

Interestingly, this has already been done with [Alertmanager](https://github.com/prometheus/alertmanager/blob/master/config/config.go#L267-L277) but not with [Prometheus](https://github.com/prometheus/prometheus/blob/main/config/config.go#L187-L195).

Please let me know if this is reasonable, and if so, if you'd like me to add some marshaling testware.
/cc @roidelapluie

### CODE REVIEW

**Comment by roidelapluie** [comment]
Please add the marshal json to hide the secrets.

**Comment by owen-d** [comment]
Before I get fancy and try to align the underlying json/yaml {un,}marshaling code, I pushed up some support for synchronizing secret, url, and http config validations.

File: config/config.go [hunk 1]
```diff
@@ -35,6 +40,11 @@ func (s *Secret) UnmarshalYAML(unmarshal func(interface{}) error) error {
 	return unmarshal((*plain)(s))
 }
 
+// MarshalJSON implements the json.Marshaler interface for Secret.
+func (s Secret) MarshalJSON() ([]byte, error) {
+	return json.Marshal(secretToken)
```
[the remark this answers is not in the document]
Comment (owen-d) [reply config/config.go:45]: I originally copied this from [Alertmanager](https://github.com/prometheus/alertmanager/blob/master/config/config.go#L66-L69) (although these types should probably be defined solely in common and imported there) and wanted to keep parity with the implementation there.

As for returning nil, we cannot actually do this as it's unsupported for the `MarshalJSON` method, see [example](https://play.golang.org/p/6RZT6VsM-4g)

I can, however, replace it with something like the following if you think it's better to show `""` for an empty secret when the field is not specified as `omitempty`.
```
// MarshalJSON implements the json.Marshaler interface for Secret.
func (s Secret) MarshalJSON() ([]byte, error) {
	if len(s) == 0 {
		return json.Marshal("")
	}
	return json.Marshal(secretToken)
}
```
How would you like me to proceed?

[the remark this answers is not in the document]
Comment (owen-d) [reply config/config.go:48]: Sounds good.

File: config/http_config_test.go [hunk 2]
```diff
@@ -36,6 +37,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/stretchr/testify/require"
```
Comment (roidelapluie) [inline config/http_config_test.go:40]: let's not depend on testify here, even for tests. Thanks.

**Comment by roidelapluie** [comment]
Thanks!

**Pull request merged**